eulerpi/examples/heat/heat.py

```diff
@@ -104,17 +105,31 @@ def forward(cls, param: np.ndarray) -> np.ndarray:
 
         solution = cls.perform_simulation(kappa=param)
 
-        # the indices of the wanted evaluation points
-        evaluation_indices = jnp.multiply(
-            cls.evaluation_points,
-            jnp.array(solution.shape),
-        ).astype(int)
-
-        solution_at_evaluation_points = jnp.array(
-            solution[
-                evaluation_indices[:, 0],
-                evaluation_indices[:, 1],
-            ]
+        # linearly interpolate the solution at the evaluation points, use own interpolation function as jax doesn't support scipy.interpolate.interp2d and doesn't provide a 2d interpolation function
+        x = jnp.linspace(0, cls.plate_length[0], cls.num_grid_points)
+        y = jnp.linspace(0, cls.plate_length[1], cls.num_grid_points)
+        X, Y = jnp.meshgrid(x, y)
+
+        # compute the indices between which the evaluation points lie
+        x_indices = jnp.searchsorted(x, cls.evaluation_points[:, 0]) - 1
+        y_indices = jnp.searchsorted(y, cls.evaluation_points[:, 1]) - 1
+        dx = cls.plate_length[0] / cls.num_grid_points
+        dy = cls.plate_length[1] / cls.num_grid_points
+
+        # interpolate the solution at the evaluation points
+        solution_at_evaluation_points = (1 / (dx * dy)) * (
+            solution[x_indices, y_indices]
+            * (x[x_indices + 1] - cls.evaluation_points[:, 0])
+            * (y[y_indices + 1] - cls.evaluation_points[:, 1])
+            + solution[x_indices + 1, y_indices]
+            * (cls.evaluation_points[:, 0] - x[x_indices])
+            * (y[y_indices + 1] - cls.evaluation_points[:, 1])
+            + solution[x_indices, y_indices + 1]
+            * (x[x_indices + 1] - cls.evaluation_points[:, 0])
+            * (cls.evaluation_points[:, 1] - y[y_indices])
+            + solution[x_indices + 1, y_indices + 1]
+            * (cls.evaluation_points[:, 0] - x[x_indices])
+            * (cls.evaluation_points[:, 1] - y[y_indices])
```

Comment:
I checked it against https://en.wikipedia.org/wiki/Bilinear_interpolation, looks good.

Depending on your time, you could refactor this into a function `interpolate_linear(x_eval, y_eval, x_grid, y_grid, func_grid)`, put it into the file `jax_extension.py` and test it using something like:
```python
n_eval = ...

f = lambda x,y:  2.0*x - y

x_grid = ...
y_grid = ...
X_grid, Y_grid = jnp.meshgrid(x_grid, y_grid)
f_grid = f(X_grid, Y_grid)

x_eval = np.random.random_sample((n_eval,))
y_eval = np.random.random_sample(n_eval,))
interp = interpolate_linear(...)

X_eval, Y_eval = jnp.meshgrid(x_eval, y_eval)
exact = f(X_eval, Y_eval)

assert jnp.allclose(interp, exact)
```

But that might already be over-engineering. I am totally fine with the current state of the pull request.

Reply:
Alright, I think I'll put that on the "future work" list :)